src/v/archival/ntp_archiver_service.cc

```diff
@@ -1581,18 +1522,93 @@ ntp_archiver::schedule_single_upload(const upload_context& upload_ctx) {
         .max_timestamp = upload.max_timestamp,
         .delta_offset = delta,
         .ntp_revision = _rev,
-        .archiver_term = upload_ctx.archiver_term,
+        .archiver_term = archiver_term,
         .segment_term = upload.term,
         .delta_offset_end = delta_offset_next,
         .sname_format = cloud_storage::segment_name_format::v3,
         .metadata_size_hint = tx_size,
       },
       .name = upload.exposed_name, .delta = offset - base,
       .stop = ss::stop_iteration::no,
-      .upload_kind = upload_ctx.upload_kind,
+      .upload_kind = upload_kind,
     };
 }
 
+ss::future<ntp_archiver::scheduled_upload>
+ntp_archiver::maybe_schedule_single_upload(
+  skip_offset_range skip_offsets,
+  model::term_id,
+  segment_upload_kind upload_kind) {
+    const auto log_level = log_level_for_error(skip_offsets.reason);
+    vlogl(
+      _rtclog,
+      log_level,
+      "Failed to make upload candidate, skipping offset range: {}",
+      skip_offsets);
+    return ss::make_ready_future<scheduled_upload>(scheduled_upload{
+      .result = std::nullopt,
+      .inclusive_last_offset = skip_offsets.end_offset,
+      .meta = std::nullopt,
+      .name = std::nullopt,
+      .delta = std::nullopt,
+      .stop = ss::stop_iteration::yes,
+      .upload_kind = upload_kind,
+    });
+}
+
+ss::future<ntp_archiver::scheduled_upload>
+ntp_archiver::maybe_schedule_single_upload(
+  candidate_creation_error error,
+  model::term_id,
+  segment_upload_kind upload_kind) {
+    const auto log_level = log_level_for_error(error);
+    vlogl(_rtclog, log_level, "failed to make upload candidate: {}", error);
+    return ss::make_ready_future<scheduled_upload>(scheduled_upload{
+      .result = std::nullopt,
+      .inclusive_last_offset = {},
+      .meta = std::nullopt,
+      .name = std::nullopt,
+      .delta = std::nullopt,
+      .stop = ss::stop_iteration::yes,
+      .upload_kind = upload_kind,
+    });
+}
+
+ss::future<ntp_archiver::scheduled_upload>
+ntp_archiver::schedule_single_upload(const upload_context& upload_ctx) {
+    auto start_upload_offset = upload_ctx.start_offset;
+    auto last_stable_offset = upload_ctx.last_offset;
+
+    auto log = _parent.log();
+
+    candidate_creation_result candidate_result;
```

Comment:
This rings alarm. I'm surprised this is default constructible, and looks like a big foot gun. It's very easy to reach line `return maybe_schedule_single_upload(` with a default constructed result. In current implementation that would be an empty `upload_candidate_with_locks` if i read the code correctly.

Reply:
Yes, I will add `std::monostate` to the variant.

Reply:
added monostate as default and assertion that we never end up with a monostate when looking at a result.

Reply:
why not use `std::optional<std::variant>>`